reader_thread.c

```diff
@@ -0,0 +1,172 @@
+// buffer size supported by our alsa interfaces (~250ms).
+// The golang runtime/scheduler does not provide any mechanism to avoid this,
+// with the suggested solution being "install more/faster cpus"
+//
+// Workaround this by creating a high proirity C thread to read into a larger
+// buffer. The end-end latency may still be relatively poor with large amounts
+// of jitter, but at least we won't be dropping audio.
+// For simplicitly this buffer uses a fixed block size equal to the
+// period of the underlying alsa device
+//
+// In theory this could probably be lockless. However that is hard.
+// Hopefully the thunk down to C code before taking the lock in _poll
+// is sufficient to isolate us from the terrible golang scheduler latency
+
+// for pthread_setname_np
+#define _GNU_SOURCE
+#include <pthread.h>
+#include <stdbool.h>
+
+#include "reader_thread.h"
+
+struct reader_thread_state_s {
+    pthread_mutex_t mu;
+    pthread_cond_t cond;
+    snd_pcm_t *h;
+    pthread_t tid;
+    char *buf;
+    int head_offset;
+    int tail_offset;
+    int period_frames;
+    int period_bytes;
+    int buf_len;
+    bool stop;
+    bool overrun;
+    bool error;
+};
+
+const char *reader_thread_error = "no error";
+
+static void *reader_thread_loop(void *arg)
+{
+    reader_thread_state *s = (reader_thread_state *)arg;
+    // Enable realtime priority for this thread
+    struct sched_param sched_param;
+    sched_param.sched_priority = sched_get_priority_max(SCHED_FIFO);
+    pthread_setschedparam(pthread_self(), SCHED_FIFO, &sched_param);
+    pthread_setname_np(pthread_self(), "goalsa_reader");
+
+    pthread_mutex_lock(&s->mu);
+    while (!s->stop) {
+        char *ptr = s->buf + s->head_offset;
+        int next_offset = s->head_offset + s->period_bytes;
+        // The last buffer may still be in use
+        if (next_offset >= s->buf_len) {
+            next_offset = 0;
+        }
+        // If the reader just can't keep up then we may see this anyway
+        if (next_offset == s->tail_offset) {
+            s->overrun = true;
+        }
+        pthread_mutex_unlock(&s->mu);
+        int rc = snd_pcm_readi(s->h, ptr, s->period_frames);
+        pthread_mutex_lock(&s->mu);
+        if (rc == -EPIPE) {
+            fprintf(stderr, "realtime alsa overrun");
+            s->overrun = true;
+            snd_pcm_prepare(s->h);
+        } else if (rc < 0) {
+            reader_thread_error = "snd_pcm_readi";
+            s->error = true;
+            s->stop = true;
+        } else if (s->overrun) {
+            // Drop data while waitng for _poll to clear overrun
+        } else {
+            if (s->head_offset == s->tail_offset) {
+                pthread_cond_signal(&s->cond);
+            }
+            s->head_offset = next_offset;
+        }
+    }
+    pthread_mutex_unlock(&s->mu);
+
+    return NULL;
+}
+
+// Returns 0 on success, 1 on overrun, -1 on error.
+// Also sets *buf_p on success;
```

Comment:
Just `*buf`?
```suggestion
// Also sets *buf on success;
```

Reply:
I changed my mind on this API half way through implementation. Comment updated.